File: Python/Problems/ProblemSet-1/Problem1-3.py

# Problem 3:

# Given the head of singly linked list,
#  find middle node of the linked list.
# If the number of nodes is odd, return the middle node.
# If the number of nodes is even,
#  there are two middle nodes, so return the second middle node.
# Example:
# Input: 1->2->3->4->5->NULL
# Output: 3 
# Explanation: There are 5 nodes in the linked list
#  and there is one middle node whose value is 3.


# Input: 10->20->30->40->50->60
# Output: 40
# Explanation: There are 6 nodes in the linked list,
#  so we have two middle nodes: 30 and 40,
#  but we will return the second middle node which is 40.


class Node: 
    def __init__(self, val = None):
        self.val = val
        self.next = None




class Linkedlist:
    def __init__(self):
        self.head = None

    def push(self, val):
        new_node = Node(val)

        if self.head is None:
            self.head = new_node
            return 
        
        last = self.head
        while last.next is not None:
            last = last.next
        last.next = new_node



        
    def middle(self):
        if self.head is None:
            raise Exception("No elements in the list")

        slow = self.head
        fast = self.head 
        while fast is not None and fast.next is not None:
            slow = slow.next  # 2, 3, 4
            fast = fast.next.next  # 3, 5, None
   
        return slow.val


  

    def __str__(self):
        ret_str = '['
        temp = self.head
        while temp is not None:
            ret_str += str(temp.val) + ', '
            temp = temp.next

        ret_str = ret_str.rstrip(', ')
        ret_str += ']'
        return ret_str


l = Linkedlist()
for i in range(1, 10): 
    l.push(i)
print(l)
print("Middle of list (odd): ",l.middle())
l.push(6)
print(l)
print("Middle (even) {Right one}:",l.middle())

File: Cpp/Problems/ProblemSets/Problem1-3.cpp

#include <iostream>
using namespace std;
// int main()

class Node
{
public:
    int val;
    Node *next;

    Node(int v)
    {
        val = v;
        next = nullptr;
    }
};

class LinkedList
{
public:
    Node *head;
    LinkedList()
    {
        head = nullptr;
    }

    void Push(int val)
    {
        Node *new_node = new Node(val);
        if (head == nullptr)
        {
            head = new_node;
            return;
        }

        Node *last = head;
        while (last->next != nullptr)
        {
            last = last->next;
        }

        last->next = new_node;
    }

    string toString()
    {
        string ret_str = "[";
        Node *temp = head;
        while (temp != nullptr)
        {
            ret_str += to_string(temp->val) + ", ";
            temp = temp->next;
        }

        if (ret_str.size() > 1)
        {
            ret_str.pop_back();
            ret_str.pop_back();
        }

        ret_str += "]";
        return ret_str;
    }

    int middle_right_element()
    {
        if (head == nullptr)
        {
            runtime_error("List is empty");
        }
        Node *slow = head;
        Node *fast = head;
        while (fast != nullptr && fast->next != nullptr)
        {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow->val;
    }
};

int main()
{

    // Here is a simple concept which i used in for loop:
    //     arr = {1, 2, 4, 5, 7, 8}
    // int = 4 bytes (normally)

    // sizeof(arr) = 6 * 4 = 24 bytes
    // sizeof(arr[0]) = 4 bytes

    // size = 24 / 4 = 6

    LinkedList l;
    int arr[] = {1, 3, 5, 7, 9, 2, 4, 6};
    int size = sizeof(arr) / sizeof(arr[0]);
    for (int i = 0; i < size; i++)
    {
        l.Push(arr[i]);
    }
    cout << l.toString() << endl;
    cout << "Middle right element: " << l.middle_right_element() << endl;

    return 0;
}